### PR TITLE
Add checklists + restructure instructions   

### LINKED ISSUE #111: Create a GH issue template with a checklist for releases

### PR DESCRIPTION
**Pull request by carlisia**
Closes #111.

What: This PR's big goal is to decouple these concepts:
-  procedure documentation
- listing of things to be done
- sequencing of things to be done

How:
- adds 1 GH issue templates with a list of todos for the release
- adds a `PROCEDURE.md` file that:
	- documents relevant operations related to a release
	- includes a short version of a ToC at the top specifying the specific order in which to perform operations
	- adds an **Exceptions** section (for operations that have exceptions) so when going thru the instructions in order, it is clear what is to be skipped 
- adds a ~`PHASES.md`~ `TIMELINE.md` file that:
	- functions as an entry point for where, when and how to get started with the release work
	- groups repos and operations for a release into separate checkpoints in the release timeline, much in the way that was in the original doc
	- includes links to only relevant sections in the (Procedure) documentation related to a particular checkpoiint

- removes the `INSTRUCTIONS.md` file, since the info in there has been split into the above two new files; didn't reuse the name "instructions" because both new files contain instructions, so named them more precisely to reflect the content of each.

There were a lot of minor edits to the documentation.

Minor changes worth noting:
- spruced up the  `README.me`. 
	- removed the table of repos, since they are all listed (per phase) in the `PHASES.md` file. If it is useful to have that unified table of repos up on the front page, I can put it back. ❓
- changed the PR used as reference to add/remove leads (in the `PROCEDURES.md` file) since the existing one being used didn't update the `OWNERS_ALIASES` file
- I don't know if the value for the `assignees` field in the templates will work. Using `knative/knative-release-leads` failed the regex validation, so I used only `knative-release-leads` to test. Either it will work, or be ignored.
                                        
Thanks to @nader-ziada and @psschwei for the prior improvements!

---
updated on 6/8

### CODE REVIEW

**Comment by psschwei** [comment]
@carlisia is this still WIP? (looking forward to giving this a go on the next release cycle!)

**Comment by carlisia** [comment]
Hey @psschwei yes, still wip, will update shortly!

**Comment by carlisia** [comment]
Pinging ppl who might be interested in reviewing this change. Please don't all feel required to review 🤓 

- current @knative/knative-release-leads 
- incoming release leads: @psschwei / @evankanderson 
- prior release leads: @nader-ziada / @kvmware / @xtreme-sameer-vohra / @dprotaso

**Review by psschwei — COMMENTED** [review]
Overall, I think this is great and will make the release process _much_ easier going forward! 

One general question: what do you think about adding direct links to the procedures from the task items? For example, when starting on the "updating the releasability defaults" task in template 2, I think it'd be nice to be able to click straight to [that section](https://github.com/knative/release/blob/87d6a11192a86def0368c8087b4652b4456492db/PROCEDURES.md#updating-the-releasability-defaults) rather than going through the phases doc to get there.

**Review by evankanderson — COMMENTED** [review]
A few comments, but lgtm

File: .github/ISSUE_TEMPLATE/9-post-release.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    description: |
+      - This release checklist tracks the work to be done post release.
+      - View [documentation](https://github.com/knative/release/blob/main/PHASES.md#post-release).
+    options:
+      - label: All items on the "T-minus zero first last batch checklist" have been checked off
+        required: true
+      - label: Homebrew tap has been updated
+      - label: The client plugin's Homebrew repo has been updated
+      - label: Release schedule has been updated
+      - label: An announcement was made in the **#general** and **#operation** Slack channels that the new Knative release is out
```
[the remark this answers is not in the document]
Comment (carlisia) [reply .github/ISSUE_TEMPLATE/9-post-release.yaml:20]: Good question. It wasn't a task in the original doc. I'll just add "blog post published" to this one, and another one something like "blog post drat started" to one of the previous phases. Additional instructions for those tasks if needed might be added in a separate PR.

File: .github/ISSUE_TEMPLATE/8-t-minus-0-last-batch.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      - Please only mark a repo as done after it has been verified the release has been published. Release notes can be done later.
+      - View [documentation](https://github.com/knative/release/blob/main/PHASES.md#t-minus-zero-last-batch).
+    options:
+      - label: All items on the "T-minus zero third batch checklist" have been checked off
+        required: true
+      - label: knative.dev/eventing-kafka-broker
+      - label: knative.dev/eventing-autoscaler-keda
+      - label: knative.dev/kn-plugin-admin
+      - label: knative.dev/kn-plugin-event
+      - label: knative.dev/kn-plugin-source-kafka
```
[the remark this answers is not in the document]
Comment (psschwei) [reply .github/ISSUE_TEMPLATE/8-t-minus-0-last-batch.yaml:22]: Looks like the func plugin isn't quite on the release train yet, at least judging by versioning... last release a couple of days ago was v0.24

File: .github/ISSUE_TEMPLATE/7-t-minus-0-third-batch.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  attributes:
+    label: Please change the title to include the current release.
+    description: |
+      - This release checklist tracks the third batch of core repos to be released.
+      - Please only mark a repo as done after it has been verified the release has been published. Release notes can be done later.
+      - View [documentation](https://github.com/knative/release/blob/main/PHASES.md#t-minus-zero-third-batch).
+    options:
+      - label: All items on the "T-minus zero second batch checklistt" have been checked off
+        required: true
+      - label: knative.dev/client
```
[the remark this answers is not in the document]
Comment (carlisia) [reply .github/ISSUE_TEMPLATE/7-t-minus-0-third-batch.yaml:18]: I sorted mindlessly grouped them following the original grouping. See screenshot below.

But now that you pointed this out... you are correct. This third batch requires both `eventing` and `serving` to be done, and the second batch only requires `eventing`.  But since both `eventing` and serving` are on batch 1, then this can roll into batch 2.

Or I can organize it some other way. Thinking/suggestions accepted.

CleanShot 2022-06-03 at 14 39 51@2x

File: PHASES.md [hunk 4]
```diff
@@ -1,155 +1,130 @@
...
+- Please do not open a new checklist before being done with the one that comes before
 
-## First week of the rotation
+## Phases
+   - [T-minus 30 days](#t-minus-30-days)
+   - [T-minus 14 days](#t-minus-14-days)
+   - [T-minus 7 days](#t-minus-7-days)
+   - [T-minus 1 day](#t-minus-1-day)
+   - [T-minus zero - day of the release](#t-minus-zero---day-of-release)
+    - [T-minus zero first batch](#t-minus-zero-first-batch)
```
[the remark this answers is not in the document]
Comment (carlisia) [reply PHASES.md:23]: I think the tradeoff here is precision vs readability and usability for humans. Given all the manual checks that need to be done, might not be a bad idea to have additional grouping (phases) that can be tackled and verified in smaller pieces, even if they could be combined into one bigger one.

File: PROCEDURES.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+### Aligning dependencies
+Each repo needs to be successfully updated to use the latest version of all shared dependencies **before** its release branch is cut.
+
+In order to align the `knative.dev` dependencies, `knative-sandbox/knobots` automation will run PRs like this [Upgrade to latest dependencies PR](https://github.com/knative/eventing/pull/4713) for each repo, executing the command `./hack/update-deps.sh --upgrade --release 0.20` and committing all the content. Note: `buoy check`, which is invoked in the script, will fail if the dependencies are not yet ready.
+
+- If there is no "Upgrade to latest dependencies" PR open, manually trigger the generation of these PRs starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. nUse this only for releases.` to **true**, because in some cases there are no code changes in the vendor.
```
[the remark this answers is not in the document]
Comment (carlisia) [reply PROCEDURES.md:40]: Yes, thank you for that.

**Comment by carlisia** [comment]
> Overall, I think this is great and will make the release process _much_ easier going forward!
> 
> One general question: what do you think about adding direct links to the procedures from the task items? For example, when starting on the "updating the releasability defaults" task in template 2, I think it'd be nice to be able to click straight to [that section](https://github.com/knative/release/blob/87d6a11192a86def0368c8087b4652b4456492db/PROCEDURES.md#updating-the-releasability-defaults) rather than going through the phases doc to get there.

Hm... that is how I started doing it. I decided to go this way to avoid having a bunch of links in the checklist. Having only 1 entry point to the documentation for each checklist and having it match to the corresponding phase made it feel like there is some uniform convention. And this avoids the following nightmare scenario: someone changes anything to do with the sections of that Procedures file, and now a linke  broken or not as relevant. It could be hard to remember, and a pain to also have to, change it in the checklist.

File: PHASES.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (carlisia) [reply PHASES.md:23]: I **think** that was the spirit of the original docs. But I also see a benefit of the "rolling horizon" idea, which would give the green light for work to start on the "next thing", rather than holding up until a batch is done.

Hm... now I don't know. 🙂

**Comment by dprotaso** [comment]
> adds 9 GH issue templates

My preference is to consolidate the templates into a single one.

Some advantages in my mind:
1. There would be a stable link for the entire release process - so anyone could visit that link and see the overall progress that's been made
2. More ergonomic - single issue vs. opening/closing 9 issues.

**Comment by psschwei** [comment]
>  scenario: someone changes anything to do with the sections of that Procedures file, and now a linke broken or not as relevant. It could be hard to remember, and a pain to also have to, change it in the checklist.

That's funny, my initial worry was the opposite, where someone changes something in the Procedures but doesn't update the issue and/or the Phases doc. If it's six in one and half dozen in the other, may as well go with the bird in hand (I thought about trying to work one more bird allusion in there, but couldn't quite bring myself to cross that road...)

> My preference is to consolidate the templates into a single one.

I think I agree with Dave on this, though I don't think Github wanted me to tell you about it:

image

File: .github/ISSUE_TEMPLATE/7-t-minus-0-third-batch.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (carlisia) [reply .github/ISSUE_TEMPLATE/7-t-minus-0-third-batch.yaml]: I "fixed" this by mostly reverting to the way the original way things were laid out.

**Comment by carlisia** [comment]
Note: there's no instruction for how to put together a blog post for the release. I added an issue for that to be added later: [Add instructions for how to put together a blog post for a new release · Issue #137 · knative/release](https://github.com/knative/release/issues/137).

File: PROCEDURES.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+## Dependency check
+tl;dr: check that all the dependencies for the repo are up-to-date and aligned with the relase version.
+
+### Exceptions
+Repos that don't have dependencies naturally don't need a dependency check and this step can be skipped for those. Currently, `knative.dev/hack` is the only `knative.dev` repo that does not have any dependencies.
+
+### Aligning dependencies
+Each repo needs to be successfully updated to use the latest version of all shared dependencies **before** its release branch is cut.
+
+In order to align the `knative.dev` dependencies, `knative-sandbox/knobots` automation will run PRs like this [Upgrade to latest dependencies PR](https://github.com/knative/eventing/pull/4713) for each repo, executing the command `./hack/update-deps.sh --upgrade --release 0.20` and committing all the content. Note: `buoy check`, which is invoked in the script, will fail if the dependencies are not yet ready.
```
Comment (psschwei) [inline PROCEDURES.md:38]:
```suggestion
In order to align the `knative.dev` dependencies, `knative-sandbox/knobots` automation will run "Upgrade to latest dependencies PRs ([example](https://github.com/knative/eventing/pull/4713) for each repo, executing the command `./hack/update-deps.sh --upgrade --release 0.20` and committing all the content. Note: `buoy check`, which is invoked in the script, will fail if the dependencies are not yet ready.
```

File: PROCEDURES.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+- If there is no "Upgrade to latest dependencies" PR open, the update PR might already have been merged. If this is not the case, manually trigger the generation of this PR starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. nUse this only for releases.` to **true**, because in some cases there are no code changes in the vendor.
```
Comment (psschwei) [inline PROCEDURES.md:40]:
```suggestion
- If there is no "Upgrade to latest dependencies" PR open, the update PR might already have been merged. If this is not the case, manually trigger the generation of this PR starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. Use this only for releases.` to **true**, because in some cases there are no code changes in the vendor.
```

File: PROCEDURES.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+### Verifying releasability
+An automatic repo releasability test is executed periodically and the results are posted on the corresponding Slack release channel. If the dependencies are properly aligned, the releasability test will pass.
+
+A test can be re-run by manually running the [Releasability workflow](https://github.com/knative/serving/actions?query=workflow%3AReleasability).
+
+If the releasability test reports a NO-GO on a repo where it was previously passing, probably a new PR merge introduced a dependency misalignment. Whatever the reason, if the releasability test is failing, it is necessary to start with the first step in the [Releasing a repository](#releasing-a-repository) process to get this test to a passing state.
+
+## Cut a branch
+tl;dr: cut a release branch from `main`.
+### Exceptions
```
Comment (psschwei) [inline PROCEDURES.md:69]:
```suggestion
tl;dr: cut a release branch from `main`.

### Exceptions
```

File: PROCEDURES.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+For some repositories some extra manual validation and updates need to be performed before the release branch is cut:
+
+- `knative/client`
+  - (optional) Check that [CHANGELOG.adoc](https://github.com/knative/client/blob/main/CHANGELOG.adoc) contains a section about the release, i.e. the top-level "(Unreleased)" section should be changed to point to the upcoming release version and number. It's not critical if the changelog is aligned after the release in retrospective.
+  - If the validation fails, the fix should be trivial and could be either performed by the release leads or the client team.
+
+- `knative-sandbox/kn-plugin-quickstart`
+  - Update the version numbers of Serving / Kourier / Eventing in [pkg/install/install.go](https://github.com/knative-sandbox/kn-plugin-quickstart/blob/main/pkg/install/install.go#L25-L27) so that the plugin will use the just-released versions.
```
Comment (psschwei) [inline PROCEDURES.md:77]:
```suggestion
  - Update the version numbers of Serving / Kourier / Eventing in [pkg/install/install.go](https://github.com/knative-sandbox/kn-plugin-quickstart/blob/4e57ed10eaac8882b0a81b2861ddaee37bfcd6f9/pkg/install/install.go#L25-L27) so that the plugin will use the just-released versions.
```

Probably better to pin to a specific commit, as the line references might change if updates are made (this one's on me... I added this line a while back and really should've used an explicit reference :facepalm: )

File: PROCEDURES.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+### Cutting a branch
+Cutting a `release-x.y` branch can be done by using the GitHub UI:
+
+1. Click on the branch selection box at the top level page of the repository.
+
+  ![Click the branch selection box](images/github-branch.png)
+
+2. Search for the correct `release-x.y` branch name for the release.
```
Comment (psschwei) [inline PROCEDURES.md:86]:
```suggestion
1. Search for the correct `release-x.y` branch name for the release.
```

File: PROCEDURES.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+- Copy `kn.rb` to the `kn@${PREV_RELEASE}.rb` with `$PREV_RELEASE` to be replace with the latest release (e.g. `0.19`).
+- In `kn@${PREV_RELEASE}.rb` replace `class Kn` with `class KnAT${PREV_RELEASE_DIGITS}`, e.g `class KnAT019` for an previous release `0.19`.
+- In `kn.rb`:
+  - Replace the old version number in `v` with the released version (e.g. `v = "v0.20.0"`)
+  - Replace the `sha256` checksums with the values from the [client release page](https://github.com/knative/client/releases). The checksums have been released too (e.g. [checksums.txt](https://github.com/knative/client/releases/download/v0.22.0/checksums.txt))
+
+✅ Open a PR and merge the changes. Prow is not enabled for the homebrew repo, so the merge needs to be performed manually.
+
+### homebrew-kn-plugins
```
Comment (psschwei) [inline PROCEDURES.md:167]: I just added a script to automate the version bump for the homebrew plugin formulas: https://github.com/knative-sandbox/homebrew-kn-plugins/blob/main/hack/update-versions.sh  Going to add one for the client formula also... hopefully makes this process a little easier

**Review by psschwei — COMMENTED** [review]
A couple of nits, but otherwise LGTM

**Comment by carlisia** [comment]
Thanks for the very helpful review @psschwei! I believe I hit them all.

Ready for another round.

**Review by psschwei — COMMENTED** [review]
/lgtm
/approve
/hold
for others to comment

**Review by gabo1208 — APPROVED** [review]

**Review by evankanderson — COMMENTED** [review]
/lgtm
/approve
/hold cancel
